README
CMConvert - CacheMate import file converter

This program is used to convert EasyGPS XML file formats (LOC and GPX) to 
a format that can be installed onto a Palm OS device and imported into 
CacheMate.  Options are also available to list waypoints contained in an 
XML file, and selectively convert specified waypoints.

CMConvert requires the Expat XML parser library.  This library is 
available at:

	http://expat.sourceforge.net/

The following libraries are also required for ZIP file support.  This is 
an optional feature, and will be disabled if these libraries cannot be 
found.

	zziplib		http://zziplib.sourceforge.net/
	zlib		http://www.gzip.org/zlib/

If ZIP file support is enabled, and such a file is given as an input file,
CMConvert will search it (by extension) for a GPX file, and take its input
from the first one it finds.  Subsequent GPX files or other files in the 
ZIP file are ignored.

For instructions on compiling and installation of CMConvert, please read 
the INSTALL file, located in the top level distribution directory.  For 
information on using the program, refer to the included man page.

-----

Copyright 2003-2005 Brian Smith

This file is part of CMConvert.

CMConvert is free software; you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation; either version 2, or (at your option)
any later version.

CMConvert is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with CMConvert; see the file COPYING.  If not, write to
the Free Software Foundation, Inc., 59 Temple Place - Suite 330,
Boston, MA 02111-1307, USA.



ChangeLog
unrelease
	* Fixed build on UTF-8 environment
	* Produce working pdp on Apple M1

2010-05-12	Version 1.9.6

	* Added support for Groundspeak's cache attributes
	* Removed option for long description text - it is now always on

2006-10-14	Version 1.9.5

	* Added support for <cmt> GPX element

2006-07-14

	* Added support for </p> tag (newline, since one is implied)

2006-05-05	Version 1.9.4

	* Better handling of HTML entities
	* Strip "Waypoint|" from beginning of waypoint type, if present
	* All GPX files are now read from a given ZIP file, instead of 
	  only the first

2006-03-06	Version 1.9.3

	* Fixed handling of extended ASCII characters in non-UTF8 files
	* Fixed handling of Navicache files

2005-06-08	Version 1.9.2

	* Added blank line before new description fields that aren't 
	  supported by CM2GPX, to keep from confusing its parser.

2005-06-05	Version 1.9.1

	* Added version line to syntax display
	* Added extra newline at end of took/left notes template

2005-04-01

	* Added support for Geocaching.com.au GPX extensions

2005-03-03	Version 1.9.0

	* Improved efficiency of XML field lookups
	* Added intelligent handling of GPX waypoint extensions.  Only the 
	  extension schema needs to be defined in a map, and the parser 
	  searches for these based on namespace URI and GPX file version.

2005-03-02

	* Added support for TerraCaching.com GPX files
	* Improved handling of incorrectly escaped HTML characters
	* Improved handling of HTML tags that don't add whitespace
	* Added "image" tag support (as opposed to "img" tags... invalid, 
	  but still used and apparently work in some browsers)
	* Changed generated HTML to use bulleted lists, to work with 
	  certain programs where definition lists didn't work so well
	* Added sorting by waypoint to generated HTML

2004-11-29	Version 1.8.7

	* Made the multi-word options add to, as opposed to replace, 
	  filters specified earlier in the command line.  This does not 
	  affect the radius filter.
	* Added support for "filter files", containing name=value pairs 
	  corresponding to string filter options.

2004-11-09

	* Updated auto-configure of integer types

2004-09-12	Version 1.8.6

	* Added "excl" command line parameter, for excluding waypoint IDs
	  from the converted file

2004-09-03

	* Added preliminary support for GPX 1.1 (no Groundspeak 
	  extensions, since they haven't been merged in yet)
	* Added support for multiple URLs per waypoint (first URL that
	  looks like a geocaching.com cache page is taken for the cache
	  URL... any others are taken as added links for the generated
	  HTML file)
	* Move long waypoint descriptions to description field (instead of 
	  name field) in the case of non-geocache GPX files without
	  Groundspeak extensions, and copy waypoint name to name field in 
	  those cases

2004-04-27	Version 1.8.5

	* Added use of file timestamp if GPX timestamp not detected

2004-04-06	Version 1.8.4

	* Added option to semi-intelligently strip quotes from cache names

2004-03-21	Version 1.8.3

	* Added proper duplicate resolution, for GPX files that contain 
	  timestamps
	* Added to documentation to explain just how duplicate record
	  resolution works in CMConvert, partly to explain the
	  "duplicates" that some people were seeing before

2004-03-05	Version 1.8.2

	* Added option to add cache active/inactive status to description
	* Added filters for active/inactive caches
	* Improved handling of incorrect list HTML (<li> tag with no
	  matching </li>).  Implicit </li> now assumed when it is missing
	  and another <li> tag is encountered.

2004-02-18	Version 1.8.1

	* Fixed Win32 problem with PDB file open call (another M$ 
	  misinterpretation of something pretty standard)

2004-02-05

	* Fixed Win32 portability issue with path separator character

2004-01-26	Version 1.8.0

	* Added warning for geocache GPX files with empty descriptions

2004-01-23

	* Moved byte order test from runtime to configure script

2004-01-22

	* Added option to generate an HTML file with any link or image 
	  URLs gleaned from HTML cache descriptions
	* Added support for statute miles to radius filter

2004-01-14

	* Added files (minus 3rd-party libraries and includes) for Win32 
	  build
	* Cleaned up a few things where Visual C++ gave warnings 
	  (unreferenced or double-defined variables)

2004-01-10	Version 1.7.2

	* Changed radius filter distance factors to be more accurate

2004-01-07	Version 1.7.1

	* Added an alert to show if descriptions were truncated for any
	  record (disabled if quiet mode specified)

2004-01-02	Version 1.7.0

	* Added support for reading zipped GPX files (when both zlib and 
	  zziplib are also installed)
	* Fixed a crash in the XML parser if encoding name is missing

2003-12-28	Version 1.6.3

	* Added lookup table for redundant Unicode characters, to replace 
	  those with high ASCII characters in the Palm OS standard fonts

2003-12-27	Version 1.6.2

	* Additional cleanup of UTF-8 decoder

2003-12-22	Version 1.6.1

	* Added proper UTF-8 decoding of text (since Expat apparently 
	  doesn't do it)
	* Removed stripping of high ASCII characters from XML input

2003-12-17	Version 1.6.0

	* Added some more portability checks with autoconf (so, for one,
	  we won't try using the functions needed to reformat dates if 
	  they aren't available and, instead, will just kill that feature)
	* Added --with-expat-dir to configure script, to allow specifying
	  where Expat is installed if it isn't anywhere in the normal 
	  search paths

2003-12-16

	* Added radius filters, for distance from a set of coordinates or 
	  waypoint ID
	* Dates are now reformatted to match user locale settings

2003-12-15

	* Added -t option to auto-include items taken/left notes template
	* Added HTML tag stripping for cache hints and logs

2003-10-17	Version 1.5.0

	* Database name based on output file name
	* Slight improvements to HTML whitespace condensing

2003-10-15

	* Added better handling for table-related tags


src/parser.h
/*
    Copyright 2003-2010 Brian Smith (brian@smittyware.com)
    This file is part of CMConvert.
    
    CMConvert is free software; you can redistribute it and/or modify   
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation; either version 2 of the License, or
    (at your option) any later version.
    
    CMConvert is distributed in the hope that it will be useful,  
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.
    
    You should have received a copy of the GNU General Public License
    along with CMConvert; if not, write to the Free Software
    Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA 02111-1307 USA
*/

#ifndef _PARSER_H_INCLUDED_
#define _PARSER_H_INCLUDED_

#define CHUNK_SIZE 8192
#define MAX_LOGS_SIZE 8192

#include "wplist.h"

// Field indices
#define FLD_NAME	0
#define FLD_WAYPOINT	1
#define FLD_COORD	2
#define FLD_DESC	3
#define FLD_HINTS	4
#define FLD_TYPE	5
#define FLD_NOTES	6
#define FLD_TERRAIN	7
#define FLD_DIFFICULTY	8
#define FLD_LOGS	9
#define FLD_NAME2	10
#define FLD_TYPE2	11
#define FLD_LAT		12
#define FLD_LON		13
#define FLD_SHRT_DESC	14
#define FLD_LONG_DESC	15
#define FLD_DATETIME	16
#define FLD_OWNER	17
#define FLD_STATE	18
#define FLD_COUNTRY	19
#define FLD_CONTAINER	20
#define FLD_LOG_DATE	21
#define FLD_LOG_TYPE	22
#define FLD_LOG_FINDER	23
#define FLD_LOG_TEXT	24
#define FLD_LOG_ENC	25
#define FLD_LOG_LAT	26
#define FLD_LOG_LON	27
#define FLD_BUGS	28
#define FLD_SYMBOL	29
#define FLD_URL		30
#define FLD_LINKS	31
#define FLD_STATUS	32
#define FLD_CAMO_DIFF	33
#define FLD_LOCALE	34
#define FLD_COMMENT	35
#define FLD_CACHE_ATTR	36

#define MAX_END_FIELDS	10
#define MAX_MID_FIELDS	37

// XPath-to-field mapping
typedef struct
{
	const char *szName;
	int nField;
	int32_t nFlags;
} stPathMap;

// NamespaceURL-to-PathMap mapping
typedef struct
{
	const char *szURI;
	int bPrefix;
	const char *szExtElem;
	stPathMap *pMap;
} stExtMap;

class IXMLReader;

class CXMLParser
{
public:
	CXMLParser();

	CWPList *m_pList;

	int m_bLocWarning;
	int m_bEmptyDesc;
	string m_sFileTS;

	// Parser options from command line
	int m_bLocation;
	int m_bContainer;
	int m_bOwner;
	int m_bDate;
	int m_bShowBugs;
	int m_bDecodeHints;
	int m_nMaxLogs;
	int m_nMaxDesc;
	int m_bLogTemplate;
	int m_bQuiet;
	int m_bCacheStatus;
	int m_bStripNameQuotes;

	int ParseFile(string sPath, IXMLReader *pReader);

private:
	string m_sCurTag, m_sCurData;
	char m_buf[CHUNK_SIZE];
	string m_sRecord;
	string m_sCurPath;
	string m_sFields[MAX_MID_FIELDS];
	int m_nCurLogs;
	int m_bHasBugs;
	int m_bNonCacheFile;
	int m_bCacheActive;
	int m_bLongDesc;
	double m_dGpxVer;
	stExtMap *m_pExtensions;
	int m_bInclAttr;
	int m_bHtmlFlag;
	string m_sExtBase;

	static void HandleElemStart(void *data, const char *el, const char 
		**attr);
	static void HandleElemEnd(void *data, const char *el);
	static void HandleCharData(void *data, const char *s, int len);

	static void GetAttribute(const char **attr, string sName,
		string &sVal);

	static void TranslateTerraSizes(string &rSize);

	int IsWhitespace(string &rStr);
	void ClearWaypoint();
	void EncodeHints(int nField);
	void FinishWaypointRecord();
	void HTMLToText(string &sStr);
	void DecodeEntity(string &sEnt, string &sValue);
	void ConvertCoords(string &sLat, string &sLon, string &sCoord,
		double &dLat, double &dLon);
	void StripQuotes(string &rStr);
	void CompileCacheLog();
	void StripTags(string &sStr);
	void ReformatDate(string &sDate);
	void DecodeUTF8(string &sStr);
	void ExtractURL(string sTag, string sArgs);
	void StoreCacheStatus(string sAvail, string sArchived);
	void FormatFileTS(string sPath);
	void AddLinkToRecord(string sURL);
	void CheckForGPXExtensions(const char **attr);
	int GetElementInfo(string sPath, int &rField, int32_t &rFlags);

	void CleanCharRefs(char *szBuf);
	int CheckCharVal(char *pVal);
};

#endif // _PARSER_H_INCLUDED_

